### ISSUE TITLE
allow Merkle trees of larger heights than 32

### ISSUE THREAD

**Issue by aszepieniec** [Neptune-Crypto/twenty-first#186]
Right now `merkle_tree.rs` explicitly enforces the bounds

```
const MAX_NUM_NODES: usize = 1 << 32;
const MAX_NUM_LEAVES: usize = MAX_NUM_NODES / 2;
pub const MAX_TREE_HEIGHT: usize = MAX_NUM_LEAVES.ilog2() as usize;
```

which capture implicit assumptions used for casting indices between various integer types. In other words, if we drop these bound checks, we get overflow errors.

Tree height 32 is large enough for anything that is constructed locally. But in the context where a Merkle tree is constructed over the course of years or even decades, in a distributed manner, *e.g.* in a blockchain context, 32 is does not suffice. Note that for the MMR we use `u64`s as indices (as we should).

 - [ ] ~~increase bounds from 32 to 64~~
 - [ ] ~~track and fix resulting errors~~

This chore ties into another discussion that we have but did not write up yet: we are using rust's `as` cast rather liberally, and without any explicit assumptions on what `usize` can and cannot be. For the time being we expect Neptune to run on modern, high-end devices where `usize` typically has 64 bits. What happens though when Neptune is run on devices where `usize` is only 16 bits, or even 8? We probably want to a) determine if compiling to such platforms generates errors; b) if so, disallow that compilation; c) fix said errors; and d) re-enable compilation.

**Comment by elielnfinic** [comment]
Can I work on this?

**Comment by aszepieniec** [comment]
Yes, please! We would welcome a PR.

**Comment by elielnfinic** [comment]
Alright, will work on this.

**Comment by jan-ferdinand** [comment]
I'm fully in support of using an architecture independent type. :+1:

**Comment by Sword-Smith** [comment]
So I've been talking about this with @jan-ferdinand and @aszepieniec, and we've come to the conclusion that nothing needs to be changed in the `merkle_tree` module.

I'm sorry that I haven't caught this earlier, but I don't see a reason to change from `usize` indices to `u64` indices. I should have killed this issue before someone started working on it, @elielnfinic.

The reason this issue was opened is that @aszepieniec saw that Merkle Trees were being used in `neptune-core` where we also have Merkle Mountain Ranges that use `u64` for indexing into the leaves. Notice that Merkle Mountain Ranges (MMR) is a entirely different data structure from Merkle trees, as the two implementations do not share any code (other than for a few (paranoid) tests).

The reason we want to use `usize` for the public-facing interface is that `MerkleTree` uses a `Vec<Digest>` to store all nodes in the tree. *If* the public interface was changed type conversion would just happen inside of the `merkle_tree` module instead of being handled by the caller. Since the `Vec<T>` type from the standard library always uses `usize` for lookups. In other words, you would just be sweeping the restriction under the rug and not actually fixing anything if you changed the public-facing interface of `merkle_tree`.

Notice that `Vec<Digest>` is used to store all nodes:
```rust
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct MerkleTree<H>
where
    H: AlgebraicHasher,
{
    nodes: Vec<Digest>,
    _hasher: PhantomData<H>,
}
```

On the question of the size-limits: 
The limits that are in place are actually good for cross-platform support, as they ensure that no one makes a `MerkleTree` that cannot be indexed on a 32-bit architecture.

```rust
const MAX_NUM_NODES: usize = 1 << 32;
const MAX_NUM_LEAVES: usize = MAX_NUM_NODES / 2;
pub const MAX_TREE_HEIGHT: usize = MAX_NUM_LEAVES.ilog2() as usize;
```

Removing these limits would exclude 32-bit architectures from handling all instances of a `MerkleTree` that could be generated on a 64-bit machine. So the current limits just ensure that 64-bit machines don't exclude 32-bit machines.

**TL;DR**: Let's not change anything, and sorry for not killing this issue before someone started working on it.

The only thing I see that could be improved is a comment on *why* these limits are there, so we don't go down this rabbit hole again.

**Comment by elielnfinic** [comment]
Alright @Sword-Smith, thank you for clarification.
